subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java

```diff
@@ -113,43 +129,63 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, HashedClas
-                urls.add(file.toURI().toURL());
+                // Having the bridge in the classloader breaks zinc
+                if (!file.toString().contains("scala3-sbt-bridge")) {
+                    urls.add(file.toURI().toURL());
+                }
             }
-            return new URLClassLoader(urls.toArray(new URL[0]));
+            return new URLClassLoader(urls.toArray(new URL[0]), parent);
         } catch (Exception ee) {
             throw new RuntimeException(ee);
         }
     }
 
-    private static ClassLoader getCachedClassLoader(HashedClasspath classpath) {
+    private static boolean isScala3(String version) {
+        return version.startsWith("3.");
+    }
+
+    private static ClassLoader getCachedClassLoader(HashedClasspath classpath, ClassLoader parent) {
         try {
             return CLASSLOADER_CACHE.get(classpath.getHash(), new Callable<ClassLoader>() {
                 @Override
                 public ClassLoader call() throws Exception {
-                    return getClassLoader(classpath.getClasspath());
+                    return getClassLoader(classpath.getClasspath(), parent);
                 }
             });
         } catch (Exception ee) {
             throw new RuntimeException(ee);
         }
     }
 
-    private static ScalaInstance getScalaInstance(HashedClasspath hashedScalaClasspath) {
-        ClassLoader scalaClassLoader = getCachedClassLoader(hashedScalaClasspath);
-        String scalaVersion = getScalaVersion(scalaClassLoader);
+    private static ScalaInstance getScalaInstance(HashedClasspath hashedScalaClasspath) throws MalformedURLException {
         ClassPath scalaClasspath = hashedScalaClasspath.getClasspath();
-
         File libraryJar = findFile(ArtifactInfo.ScalaLibraryID, scalaClasspath);
-        File compilerJar = findFile(ArtifactInfo.ScalaCompilerID, scalaClasspath);
+        URL[] libraryUrls;
+        try {
+            File library3Jar = findFile(SCALA_3_LIBRARY_ID, scalaClasspath);
+            libraryUrls = new URL[]{library3Jar.toURI().toURL(), libraryJar.toURI().toURL()};
+        } catch (IllegalStateException e) {
+            libraryUrls = new URL[]{libraryJar.toURI().toURL()};
+        }
+        ClassLoader scalaLibraryClassLoader = new ScalaCompilerLoader(libraryUrls, xsbti.Reporter.class.getClassLoader());
+        ClassLoader scalaClassLoader = getCachedClassLoader(hashedScalaClasspath, scalaLibraryClassLoader);
+        String scalaVersion = getScalaVersion(scalaClassLoader);
+
+        File compilerJar;
+        if (isScala3(scalaVersion)) {
+            compilerJar = findFile(SCALA_3_COMPILER_ID, scalaClasspath);
+        } else {
+            compilerJar = findFile(ArtifactInfo.ScalaCompilerID, scalaClasspath);
+        }
 
         return new ScalaInstance(
             scalaVersion,
             scalaClassLoader,
-            getClassLoader(DefaultClassPath.of(libraryJar)),
+            scalaLibraryClassLoader,
```

Comment:
If I undo this change and use `getClassLoader(DefaultClassPath.of(libraryJar), null)` instead and add a null check in `getClassLoader` and similarly pass null as the second param to the call to `getCachedClassLoader` on line 175, then the test passes again.

Reply:
I can change that back for Scala 2 for sure. However, I was almost 100% sure that scala works with later jdks only since 2.12 🤔

Reply:
Well I ran tests locally with 11 and CI with 16 and they "work" with the previous setup.

Reply:
Thanks! I will get back to it tomorrow, not sure why normally it doesn't work with later JDKs, but maybe there is a difference here in Gradle that makes it work, no idea.

I will revert the change for Scala 2 to make the older tests work.

Reply:
I changed the classloader for Scala 2 to not use any changes made for Scala 3. Not 100% sure what's going on here, but that should be pretty safe to include now.